src/main/src/main/java/org/geoserver/catalog/impl/FeatureTypeInfoImpl.java

```diff
@@ -185,8 +186,73 @@ public boolean equals(Object obj) {
         if (attributes == null) {
             if (other.getAttributes() != null)
                 return false;
-        } else if (!attributes.equals(other.getAttributes()))
-            return false;
+        } else {
+            List<AttributeTypeInfo> otherAttributes = other.getAttributes();
+            if (otherAttributes == attributes)
+                return true;
+
+            ListIterator<AttributeTypeInfo> attributesIterator = attributes.listIterator();
+            ListIterator<AttributeTypeInfo> otherAttributesIterator = otherAttributes.listIterator();
+            while (attributesIterator.hasNext() && otherAttributesIterator.hasNext()) {
+                AttributeTypeInfo attr = attributesIterator.next();
+                AttributeTypeInfo otherAttr = otherAttributesIterator.next();
+
+                if (attr == null) {
+                    if (otherAttr != null)
+                        return false;
+                }
+
+                /*
+                 * Compare attributes, excluding the feature type to avoid recursion.
+                 * This should be updated whenever AttributeTypeInfoImpl.equals() changes.
+                 */
+                if (attr instanceof AttributeTypeInfoImpl && otherAttr instanceof AttributeTypeInfoImpl) {
```

Comment:
One suggestion, given that AttributeTypeInfo backlink to FeatureTypeInfo is "structural", exposed even by that interface, does it make sense to have a specialized equals method in AttributeTypeInfo whose implementation is basically the code below? Would make things clearer for other potential implementors, avoid knowing about the Impl class, and keep the code comparing AttributeTypeInfo close to it.

Reply:
Thinking about it, is there ever a case where we would want to use one over the other? Anything I can think of required already knowing something about the natures of the objects being compared (e.g. FeatureType being null).

From a code cleanliness and maintainability standpoint, I agree that adding a second equals method would be better.

As an alternative, does simply removing the check on FeatureTypeInfo from `AttributeTypeInfoImpl.equals()` (moving the contents of this PR into that equals method instead) cause a problem anywhere? It does stop is from doing an exact interface comparison, but the interface already has a build-in recursive structure.

Reply:
Thinking further, I suppose adding a second equals to AttributeTypeInfo would make sense if `FeatureTypeInfoImpl.equals()` was the main thing that made use of it. So basically the same thing that is going on here, but actually supported by the interface.

In that case, yes, seems like a good idea. I can make that change to this PR.